### PR TITLE
Add services.ConfigureOptions<T>(Type) overloads

### PR DESCRIPTION
**Pull request by HaoK**
Fixes https://github.com/aspnet/Options/issues/222

Adds a few new ways to configure options:
- ConfigureOptions(Type)
- `ConfigureOptions<TSetup>()`
- ConfigureOptions(object setupInstance)

All of these use reflection to register all of the
IConfigureOptions/IPostConfigureOptions on the type/instance

Note: I ressurrected these from what we had back in 1.0.0-beta8 (not sure about the names)

Also will be used for: https://github.com/aspnet/Options/issues/224

cc @davidfowl @ajcvickers @divega

### CODE REVIEW

File: src/Microsoft.Extensions.Options/ConfigureNamedOptions.cs [hunk 1]
```diff
@@ -9,7 +9,7 @@ namespace Microsoft.Extensions.Options
     /// Implementation of IConfigureNamedOptions.
     /// </summary>
     /// <typeparam name="TOptions"></typeparam>
-    public class ConfigureNamedOptions<TOptions> : IConfigureNamedOptions<TOptions>, IConfigureOptions<TOptions> where TOptions : class
+    public class ConfigureNamedOptions<TOptions> : IConfigureNamedOptions<TOptions> where TOptions : class
```
Comment (davidfowl) [inline src/Microsoft.Extensions.Options/ConfigureNamedOptions.cs]: Isn't this a breaking change?

Comment (HaoK) [reply src/Microsoft.Extensions.Options/ConfigureNamedOptions.cs]: IConfigureNamedOptions derives from IConfigureOptions, this is just cleanup

**Comment by davidfowl** [comment]
@HaoK should this throw if you implement both interfaces? What's the expected behavior then?

**Comment by HaoK** [comment]
So basically to do named options in a non breaking way, how it works is that IConfigureOptions are treated as IConfigureNamedOptions with name == Options.DefaultName.  

So even though they have to implement both interfaces for IConfigureNamedOptions, but the old `Configure(TOptions o)` is never called

**Comment by HaoK** [comment]
The correct implementation is always to forward like this: `Configure(TOptions o) => Configure(Options.DefaultName, o)`

**Pull request merged**